### ISSUE TITLE
Any Benchmark of Zygote?

### ISSUE THREAD

**Issue by MarisaKirisame** [FluxML/Zygote.jl#324]
Hi, I had read the paper "Dont unroll adjoint", and from my understanding, zygote had invented a new AD algorithm. However, there is also lots of AD compiler out there, and is there any head to head benchmark comparing zygote to the classical AD compiler? (e.g. tapenade, ADIFOR. even comparison to overloading approach like adept will be great). I would love to see in what benchmark does the new algorithm has advantage over the existing one.

**Comment by MikeInnes** [comment]
We are in the process of publishing up-to-date benchmarks, but here's the summary. An imperfect but basically sensible way to measure AD overhead is "operation overhead" – that is, how much time you spend manipulating data structures rather than doing numerical work when doing something basic like `a*b`. If you replace all your arrays with a `FakeArray` type that does no work, you can measure this pretty directly.

Tracing / operator overloading based ADs, like Tracker, generally have an overhead of about a microsecond. This is ok for GPUs, as long as your arrays are reasonably large, but for scalar code it's liable to make your code up to a thousand times slower. Tapenade (an SCT AD) is stellar here though – their overheads are closer to the nanosecond range.

Our benchmarks put Zygote in the 10-50 nanosecond range currently. That's not Tapenade level – it'll still make code an order of magnitude slower in the worst case – but it's still leagues ahead of most ML ADs, and is negligible for those sorts of workloads, even with small arrays.

It's possible in principle to make sure Zygote's code gets fully type-inferred, and from there we can do things like managing allocation on the stack which would get us to Tapenade-level performance. That's some way off for now though.

**Comment by MarisaKirisame** [comment]
IMO comparing to ML ADs are unfair. In ML, every one know the tensors are the bottleneck, so ppl use the most expensive but easiest technique: piggy back them onto the computation graph. Thus, they do not make contributions to AD, or claim they are SOTA AD techniques (for general AD problem).

The fairest comparison in the AD world might be lantern (Demystifying Differentiable Programming), in which they break up the trace into delimited continuation, then stage and cps to remove the shift/reset and the interpretation overhead.

Another thing is, as you had said, reducing trace is only a small part of story in AD. AD tools have investigated in other fonts, such as efficient higher order gradient, exploiting sparse matrixes, binomial checkpointing(although checkpointing in theory only help memory, it might be faster to run code with checkpointing as it reduces memory footprint, thus put more stuff on the cache), mixed forward/reverse mode to get complexity smaller then O(original_run_time * min(input_size, output_size)). See the classical book <Evaluating derivatives> for more details.

And even the 'trace time' is an imperfect computation. For example, in the book, the authors had claimed that the tape data structure take no time, given the tape is not swapped onto disk, and you have cache + superscalar. In this case the tape work time will be superscalared alongside the gradient computation. 

So, why not do end-to-end comparison? It is the fairest setup, and it is not hard to do.

**Comment by MikeInnes** [comment]
> In ML, every one know the tensors are the bottleneck ... they do not make contributions to AD

In computer vision, sure. In other areas like NLP or RL, not so much. It's neither true that the ML community think overhead is unimportant nor that they aren't contributing to AD for that reason; see plenty of other efforts like Swift for TF, Myia, Torch Script, AutoGraph, etc. Beyond AD overhead, these efforts are also important because tracing precludes further optimisations even in principle.

IMO, end to end benchmarks suites have very limited value compared to an understanding of the underlying factors that affect performance. It's a huge amount of work to both cover a reasonable set of use cases (CV, NLP, RL, optimisation, scalar code, GPU/CPU, distribution, small and large problem sizes in all of the above...) as well as making sure that you're using other peoples frameworks correctly to avoid unfairness. If you're interested in doing it yourself I'm sure we'd be happy to help with the Zygote side of it, though.

**Comment by ChrisRackauckas** [comment]
>IMO comparing to ML ADs are unfair. In ML, every one know the tensors are the bottleneck, so ppl use the most expensive but easiest technique: piggy back them onto the computation graph. Thus, they do not make contributions to AD, or claim they are SOTA AD techniques (for general AD problem).

In traditional ML like deep neural networks, sure. But what about in scientific machine learning where the adjoint pass needs to go through hand-defined scalar functions? Traditional tracing techniques are terribly slow on these kinds of codes, and I think it's great to showcase the utility of Zygote over something like TensorFlow by showcasing how it can better integrate with PDE solvers. This is a very active area of research which is demonstrating some very interesting results, and showing the community the performance advantages they can get by leaving frameworks which are not optimized to the domain is of great value to practitioners.

>So, why not do end-to-end comparison? It is the fairest setup, and it is not hard to do.

It's difficult to do real tasks. For example, with scientific machine learning it can be hard to do end-to-end comparisons in setups like TensorFlow or PyTorch because their ecosystems are rather primitive. While in Julia it can take a few hours to whip up a benchmark on an adaptive neural backwards stochastic differential equation, I know from experience that it could take some significant development time to build this non-existent library for Python (and optimize it) (it could take a year or two). Maybe when their package ecosystems mature it would be interesting to do some more end-to-end benchmarks in the SciML domain, but for now there is too much work that would have to be done by hand in order to get to a full-scale method. I personally would like to make such benchmarks but have to wait until the time is more appropriate.

But on the other hand, if you stick to where the Python libraries do exist, you've already mentioned and agree that traditional machine learning is very much bottlenecked by the cost of large tensor computations, and so that's uninteresting: everyone will be about the same, at least until you get to the size where automatic distributed parallelism gets involved.

>exploiting sparse matrixes

There already exists automated sparsity detection which integrates with Julia AD systems like Zygote. 

https://github.com/JuliaDiffEq/SparsityDetection.jl
https://github.com/JuliaDiffEq/SparseDiffTools.jl

>mixed forward/reverse mode

Zygote already handles this.

**Comment by CarloLucibello** [comment]
This was a benchmarking question, answered in-thread. Closing as stale.

_(Triaged against Zygote v0.7.10 on Julia 1.12.6.)_
